### PR TITLE
Fix DataFlowsOut for certain local function situations

### LINKED ISSUE #41600: Incorrect result is produced by DataFlowAnalysis.DataFlowsOut when local functions are involved
```

        [Fact]
        public void Test1()
        {
            var source =
@"
class P
{
    static void Main()
    {
        bool Dummy(params object[] x) {return true;}

        try {}
        catch when (TakeOutParam(out var x1) && x1 > 0)
        {
            Dummy(x1);
        }

        try {}
        catch when (Dummy(TakeOutParam(out var x14), 
                            TakeOutParam(out var x14), // 2
                            x14))
        {
            Dummy(x14);
        }

        static bool TakeOutParam(out int x) 
        {
            x = 123;
            return true;
        }
    }
}
";
            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseDll, parseOptions: TestOptions.Regular);

            var tree = compilation.SyntaxTrees.Single();
            var model = compilation.GetSemanticModel(tree);

            var node = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().First();
            Assert.Equal("TakeOutParam(out var x1)", node.ToString());
            var flow = model.AnalyzeDataFlow(node);
            Assert.Equal("x1", flow.DataFlowsOut.Single().Name);
        }
```

Expected: ```x1``` is returned

Observed: ```x1``` and ```x14``` are returned.
```
[xUnit.net 00:00:00.01] xUnit.net VSTest Adapter v2.4.1-pre.build.4059 (64-bit .NET Core 3.1.1)
[xUnit.net 00:00:02.27]   Starting:    Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests
The active test run was aborted. Reason: Test host process crashed : Process terminated. Assertion failed.
   at Microsoft.CodeAnalysis.CSharp.DataFlowsOutWalker.Analyze(CSharpCompilation compilation, Symbol member, BoundNode node, BoundNode firstInRegion, BoundNode lastInRegion, HashSet`1 unassignedVariables, ImmutableArray`1 dataFlowsIn) in F:\GitHub\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DataFlowsOutWalker.cs:line 40
   at Microsoft.CodeAnalysis.CSharp.CSharpDataFlowAnalysis.get_DataFlowsOut() in F:\GitHub\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\CSharpDataFlowAnalysis.cs:line 156
   at Microsoft.CodeAnalysis.CSharp.UnitTests.OutVarTests.Test1() in F:\GitHub\roslyn\src\Compilers\CSharp\Test\Semantic\Semantics\OutVarTests.cs:line 35332
```

No repro when ```TakeOutParam``` and ```Dummy``` are changed to be real methods vs. local functions.  Still reproduces when only ```TakeOutParam``` is converted, but ```Dummy``` remains a local function.

### PR DESCRIPTION
**Pull request by agocke**
There are two problems here, the first of which is the most important.

First, certain data flow analyses use "normalization" to adjust an existing
state variable for more variables declared after that state was created.
The default behavior is to extend that state to the "top" state, meaning
the starting state for that data flow analysis. This is usually the
right extension. However, some analysis requires extending the
variables the other way, to the bottom state. This PR adds functionality
for that customization.

The second problem is that data flow for local functions may include
variables that are not captured. This is by filtering the transfer function
to only include captured variables.

This PR is structured into two commits, each of which fix the associated
problem.

Fixes fixes #41600

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 1]
```diff
@@ -741,9 +741,18 @@ protected override void Normalize(ref LocalState state)
...
-                state.Assigned[i] = (slot > 0) &&
+
+                bool assign = (slot > 0) &&
                     state.Assigned[slot] &&
                     variableBySlot[slot].Symbol.GetTypeOrReturnType().TypeKind == TypeKind.Struct;
+
+                if (state.NormalizeToBottom)
+                {
+                    // NormalizeToBottom means new variables are assumed to be assigned (bottom state)
+                    assign |= slot == 0;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs]: >slot == 0 [](start = 30, length = 9)

Could we move this into the ```if``` condition and use ```assign = true``` here? I think that would be easier to follow.

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs [hunk 2]
```diff
@@ -4131,6 +4131,124 @@ static void Main(string[] args)
 
         #region "lambda"
 
+        [Fact]
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: >[Fact] [](start = 8, length = 6)

WorkItem?

**Comment by AlekseyTs** [comment]
Does this PR fix https://github.com/dotnet/roslyn/issues/41853 as well?

**Comment by agocke** [comment]
No, that looks like an issue specific to DataFlowsIn.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.LocalFunctions.cs [hunk 3]
```diff
@@ -112,10 +118,10 @@ private void RecordReadInLocalFunction(int slot)
             }
         }
 
-        private BitVector GetCapturedBitmask(ref BitVector state)
+        private BitVector GetCapturedBitmask()
         {
-            BitVector mask = BitVector.Empty;
-            for (int slot = 1; slot < state.Capacity; slot++)
+            BitVector mask = BitVector.AllSet(1);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.LocalFunctions.cs]: >1 [](start = 46, length = 1)

`nextVariableSlot` (so the internal array doesn't have to be reallocated).  Then in the loop,
`mask[slot] = IsCapturedInLocalFunction(slot)`

**Comment by gafter** [comment]
>         internal sealed class LocalFunctionState : AbstractLocalFunctionState

I would have expected this to override `NormalizeToBottom` to return true. Can you please explain why that's not correct?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:8552 in 64a4c6a. [](commit_id = 64a4c6a68b707e59d2d57f800dbde4586cddc68e, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs [hunk 4]
```diff
@@ -4131,6 +4131,125 @@ static void Main(string[] args)
...
+        [WorkItem(41802, "https://github.com/dotnet/roslyn/pull/41802")]
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: >41802 [](start = 18, length = 5)

41600

**Comment by gafter** [comment]
Done reviewing Iteration 3.

**Comment by agocke** [comment]
@gafter For nullable, we do a slightly different analysis since it's not possible to be flow dependent, as only one type can be reported in the semantic model. The design is to track a single "StartingState" that's Joined at every local function invocation and the starting state for the local function is the Join of every use site. I think the right behavior here is to treat undeclared variables as unassigned, since effectively they are at that point in the code.

**Pull request merged**